formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/schema/Generation.kt

```diff
@@ -0,0 +1,413 @@
+
+private val SerialDescriptor.isProtobufMessage: Boolean
+    get() = kind == StructureKind.CLASS || kind == StructureKind.OBJECT || kind == PolymorphicKind.SEALED || kind == PolymorphicKind.OPEN
+
+private val SerialDescriptor.isProtobufCollection: Boolean
+    get() = isProtobufRepeated || isProtobufMap
+
+private val SerialDescriptor.isProtobufRepeated: Boolean
+    get() = (kind == StructureKind.LIST && getElementDescriptor(0).kind != PrimitiveKind.BYTE)
+            || (kind == StructureKind.MAP && !getElementDescriptor(0).isValidMapKey)
+
+private val SerialDescriptor.isProtobufMap: Boolean
+    get() = kind == StructureKind.MAP && getElementDescriptor(0).isValidMapKey
+
+private val SerialDescriptor.isProtobufEnum: Boolean
+    get() = kind == SerialKind.ENUM
+
+private val SerialDescriptor.isValidMapKey: Boolean
+    get() = kind == PrimitiveKind.INT || kind == PrimitiveKind.LONG || kind == PrimitiveKind.BOOLEAN || kind == PrimitiveKind.STRING
+
+
+private val SerialDescriptor.messageOrEnumName: String
+    get() = serialName.substringAfterLast('.', serialName)
+
+private fun SerialDescriptor.protobufTypeName(annotations: List<Annotation> = emptyList()): String {
+    return if (isProtobufScalar) {
+        scalarTypeName(annotations)
+    } else {
+        messageOrEnumName
+    }
+}
+
+private val SerialDescriptor.protobufEnumElementName: String
+    get() = serialName.substringAfterLast('.', serialName)
+
+private fun SerialDescriptor.scalarTypeName(annotations: List<Annotation> = emptyList()): String {
+    val integerType = annotations.filterIsInstance<ProtoType>().firstOrNull()?.type ?: ProtoIntegerType.DEFAULT
+
+    if (kind == SerialKind.CONTEXTUAL) {
+        return "bytes"
+    }
+
+    if (kind is StructureKind.LIST && getElementDescriptor(0).kind == PrimitiveKind.BYTE) {
+        return "bytes"
+    }
+
+    return when (kind as PrimitiveKind) {
+        PrimitiveKind.BOOLEAN -> "bool"
+        PrimitiveKind.BYTE, PrimitiveKind.CHAR, PrimitiveKind.SHORT, PrimitiveKind.INT ->
+            when (integerType) {
+                ProtoIntegerType.DEFAULT -> "int32"
+                ProtoIntegerType.SIGNED -> "sint32"
+                ProtoIntegerType.FIXED -> "fixed32"
+            }
+        PrimitiveKind.LONG ->
+            when (integerType) {
+                ProtoIntegerType.DEFAULT -> "int64"
+                ProtoIntegerType.SIGNED -> "sint64"
+                ProtoIntegerType.FIXED -> "fixed64"
+            }
+        PrimitiveKind.FLOAT -> "float"
+        PrimitiveKind.DOUBLE -> "double"
+        PrimitiveKind.STRING -> "string"
+    }
+}
+
+private fun removeLineBreaks(text: String): String {
+    return text.replace('\n', ' ').replace('\r', ' ')
+}
+
+@SharedImmutable
+private val IDENT_REGEX = Regex("[A-Za-z][A-Za-z0-9_]*")
+
+private val String.isProtobufFullIdent: Boolean
+    get() = split('.').all { it.isProtobufIdent }
+
+private val String.isProtobufIdent: Boolean get() = matches(IDENT_REGEX)
```

Comment:
This property is used to assert that the name **is** valid identifier and throw an exception otherwise.

It's a good indicator that you need something like `checkIsValidIdentifier(name)` instead

Reply:
A single method will throw an exception with a common and not very verbose message.

Reply:
you can slightly modify it if you want to :)
Few options here: single line check-like check `checkIsValidIdentifier(name) { ...lazy error message... }` or `descriptor. checkIsValidIdentifier(name)`.

Don't really insist tho